cpp/src/sampling/temporal_sampling_impl.hpp

```diff
@@ -0,0 +1,1108 @@
+    auto next_frontier_vertex_label_spans =
+      starting_vertex_labels ? std::make_optional<std::vector<raft::device_span<label_t const>>>()
+                             : std::nullopt;
+    auto next_frontier_vertex_time_spans =
+      std::make_optional<std::vector<raft::device_span<edge_time_t const>>>();
+    size_t no_duplicates_size{0};
+    size_t has_duplicates_size{0};
+
+    auto start_offset = hop * (num_edge_types ? *num_edge_types : edge_type_t{1});
+    auto end_offset =
+      start_offset +
+      std::min(static_cast<size_t>((num_edge_types ? *num_edge_types : edge_type_t{1})),
+               fan_out.size() -
+                 hop * static_cast<size_t>(num_edge_types ? *num_edge_types : edge_type_t{1}));
+    for (size_t i = start_offset; i < end_offset; ++i) {
+      if (fan_out[i] > 0) {
+        if (!level_Ks) {
+          level_Ks = std::vector<size_t>(num_edge_types ? *num_edge_types : edge_type_t{1}, 0);
+        }
+        (*level_Ks)[i - start_offset] = fan_out[i];
+      } else if (fan_out[i] < 0) {
+        if (!gather_flags) {
+          gather_flags = std::vector<uint8_t>(num_edge_types ? *num_edge_types : edge_type_t{1},
+                                              static_cast<uint8_t>(false));
+        }
+        (*gather_flags)[i - start_offset] = static_cast<uint8_t>(true);
+      }
+    }
+
+    if (hop > 0) {
+      // It's possible for a vertex to appear in the frontier multiple times with different time
+      // stamps (after the first hop).  To handle that situation, we need to partition the vertex
+      // set.
+      std::tie(frontier_vertices_no_duplicates,
+               frontier_vertex_times_no_duplicates,
+               frontier_vertex_labels_no_duplicates,
+               frontier_vertices_has_duplicates,
+               frontier_vertex_times_has_duplicates,
+               frontier_vertex_labels_has_duplicates) =
+        temporal_partition_vertices(
+          handle,
+          raft::device_span<vertex_t const>{frontier_vertices.data(), frontier_vertices.size()},
+          raft::device_span<edge_time_t const>{frontier_vertex_times->data(),
+                                               frontier_vertex_times->size()},
+          std::make_optional(raft::device_span<label_t const>{frontier_vertex_labels->data(),
+                                                              frontier_vertex_labels->size()}));
+
+      no_duplicates_size  = frontier_vertices_no_duplicates.size();
+      has_duplicates_size = frontier_vertices_has_duplicates.size();
+
+      if constexpr (multi_gpu) {
+        no_duplicates_size = host_scalar_allreduce(
+          handle.get_comms(), no_duplicates_size, raft::comms::op_t::SUM, handle.get_stream());
+        has_duplicates_size = host_scalar_allreduce(
+          handle.get_comms(), no_duplicates_size, raft::comms::op_t::SUM, handle.get_stream());
+      }
+
+      if (no_duplicates_size > 0) {
```

Comment:
Have you considered the case `graph_view` already has an attached mask?

Reply:
In this case, you need to clear edge mask before attaching a new one. Also, edge_time_mask's values for masked-out edges won't be updated in primitive calls, so they may have undefined values.

Reply:
I think I'll add a CUGRAPH_EXPECTS and a FIXME on this one.  Fast fail at the beginning if the graph_view has an attached max, FIXME to indicate that we should support that.